felix/nftables/table.go

```diff
@@ -0,0 +1,1293 @@
+		// as insert chain/rules above.
+		ourAppendedHashes = CalculateRuleHashes(chainName+"*appends*", appendedRules, features)
+	}
+	time.Sleep(1 * time.Second)
+	offset := 0
+	for i, hash := range ourInsertedHashes {
+		allHashes[i+offset] = hash
+	}
+
+	offset = len(insertedRules) + numNonCalicoRules
+	for i, hash := range ourAppendedHashes {
+		allHashes[i+offset] = hash
+	}
+	return
+}
+
+// getHashesAndRulesFromDataplane loads the current state of our table. It parses out the hashes that we
+// add to rules and, for chains that we insert into, the full rules. The 'hashes' map contains an entry for each chain
+// in the table. Each entry is a slice containing the hashes for the rules in that table. Rules with no hashes are
+// represented by an empty string. The 'rules' map contains an entry for each non-Calico chain in the table that
+// contains inserts. It is used to generate deletes using the full rule, rather than deletes by line number, to avoid
+// race conditions on chains we don't fully control.
+func (t *nftablesTable) getHashesAndRulesFromDataplane() (hashes map[string][]string, rules map[string][]*knftables.Rule) {
+	retries := 3
+	retryDelay := 100 * time.Millisecond
+
+	// Retry a few times before we panic.  This deals with any transient errors and it prevents
+	// us from spamming a panic into the log when we're being gracefully shut down by a SIGTERM.
+	for {
+		t.onStillAlive()
+		hashes, rules, err := t.attemptToGetHashesAndRulesFromDataplane()
+		if err != nil {
+			countNumSaveErrors.Inc()
+			var stderr string
+			if ee, ok := err.(*exec.ExitError); ok {
+				stderr = string(ee.Stderr)
+			}
+			t.logCxt.WithError(err).WithField("stderr", stderr).Warn("nftables command failed")
+			if retries > 0 {
+				retries--
+				t.timeSleep(retryDelay)
+				retryDelay *= 2
+			} else {
+				t.logCxt.Panic("nftables command failed after retries")
+			}
+			continue
+		}
+
+		return hashes, rules
+	}
+}
+
+// attemptToGetHashesAndRulesFromDataplane reads nftables state and loads it into memory.
+func (t *nftablesTable) attemptToGetHashesAndRulesFromDataplane() (hashes map[string][]string, rules map[string][]*knftables.Rule, err error) {
+	startTime := t.timeNow()
+	defer func() {
+		saveDuration := t.timeNow().Sub(startTime)
+		t.peakNftablesReadTime = t.peakNftablesReadTime * 99 / 100
+		if saveDuration > t.peakNftablesReadTime {
+			t.logCxt.WithField("duration", saveDuration).Debug("Updating nftables peak duration.")
+			t.peakNftablesReadTime = saveDuration
+		}
+	}()
+
+	t.logCxt.Debug("Attmempting to get hashes and rules from nftables")
+	chains, err := t.nft.List(context.TODO(), "chain")
+	if err != nil {
+		return nil, nil, err
+	}
+	countNumSaveCalls.Inc()
+
+	hashes = make(map[string][]string)
+	rules = make(map[string][]*knftables.Rule)
+
+	for _, chain := range chains {
+		hashes[chain] = []string{}
+		rulesInChain, err := t.nft.ListRules(context.TODO(), chain)
```

Comment:
Unfortunately making this many nft calls seems to make loading the dataplane even for a pretty small ruleset take >500ms, which doesn't look good for scaling. 

We'll need to update this to be more efficient. We should be able to load the entire table in one API call, but `knftables` doesn't support that (yet, at least) so perhaps they will accept a PR upstream for this one.